### DISCUSSION TITLE
How to get the returned id?

### DISCUSSION THREAD

**Discussion by Ilqjx** [rbatis/rbatis#274] [General]
```
pub struct A {
    id: i32,
    name: String,
    map: HashMap<String, String>,
}

pub async fn save_a(a: A) {
    let res = RB.exec("insert into a(id, name) values($1, $2)",
            vec![Bson::Int32(a.id), Bson::String(a.name)]).await.unwrap();
}
```

**Comment by zhuxiujia** [comment]
> ```
> pub struct A {
>     id: i32,
>     name: String,
>     map: HashMap<String, String>,
> }
> 
> pub async fn save_a(a: A) {
>     let res = RB.exec("insert into a(id, name) values($1, $2)",
>             vec![Bson::Int32(a.id), Bson::String(a.name)]).await.unwrap();
> }
> ```

the ``` exec``` method return data have  field ```last_insert_id ```, this is database returned id

**Comment by nanokes** [comment]
Is there a reason that `last_insert_id` is of type `Option<i64>` instead of `Option<u64>`. Is there a time when `last_insert_id` is negative?

**Comment by Ilqjx** [comment]
@zhuxiujia Thanks for your reply

**Comment by zhuxiujia** [comment]
> 

i think postgresql not support last_insert_id.   This is a limitation at the database-driven do this

**Comment by Ilqjx** [comment]
@zhuxiujia I think this feature is very important, can you consider using other ways to support it?

**Comment by zhuxiujia** [comment]
> @zhuxiujia I think this feature is very important, can you consider using other ways to support it?

You can choose to generate IDS ahead of time and insert them, or use RETURNING IDS. I prefer the former. If the former is selected, use contiguous distributed ids。you can see https://github.com/rbatis/rbatis/blob/master/src/plugin/object_id.rs

**Comment by Ilqjx** [comment]
> > @zhuxiujia I think this feature is very important, can you consider using other ways to support it?
> 
> You can choose to generate IDS ahead of time and insert them, or use RETURNING IDS. I prefer the former. If the former is selected, use contiguous distributed ids。you can see https://github.com/rbatis/rbatis/blob/master/src/plugin/object_id.rs

Ok, Thanks.
